nixos/modules/services/networking/iwd.nix

```diff
@@ -4,16 +4,141 @@ with lib;
 
 let
   cfg = config.networking.wireless.iwd;
+
+  encodeFileName = pkgs.writeScript "encode" ''
+    #! ${pkgs.runtimeShell} -e
+    echo "=$(printf "$1" | ${pkgs.unixtools.xxd}/bin/xxd -pu)"
+  '';
+
+  # If a network consists of alphanumeric chars and `_` and `-`, the network will be stored
+  # as `/var/lib/iwd/<ssid>.<ext>`. In any other case a hex-encoded version of the SSID will
+  # be stored with an `=` as prefix.
+  encodeIfNeeded = file: ext: let
+    allowedChars = stringToCharacters "01234556789abcdefghijklmnopqrstuvwxyz-_";
+    skipEncoding = all (char: elem (toLower char) allowedChars) (stringToCharacters file);
+  in
+    "/var/lib/iwd/${if skipEncoding then file else "$(${encodeFileName} \"${file}\")"}.${ext}";
+
+  # Basically `types.str`, but the `check` function validates the string-length as well.
+  fixedStringLength = min: max: mkOptionType {
+    name = "str-with-length";
+    description = if min == max
+      then "string with exactly ${toString min} chars."
+      else "string which must have at least ${toString min} and at most ${toString max} chars.";
+    inherit (types.str) merge;
+    check = x: isString x && stringLength x >= min && stringLength x <= max;
+  };
+
+  # The file extension of the network-config determines the type of the network. Currently,
+  # we support public networks (`open`) and networks protected with a passphrase (`psk`).
+  # Anything else, like 802.1x networks are not supported by this module.
+  determineType = cfg: if cfg.psk != null || cfg.passphrase != null then "psk" else "open";
+
+  # Generates a name-value-pair with configurations for a SSID. For further information
+  # about the configuration options, please refer to the upstream docs:
+  #
+  # https://iwd.wiki.kernel.org/networkconfigurationsettings
+  predefinedNetworks = flip mapAttrs cfg.networks (name: cfg: {
+    extension = determineType cfg;
+
+    # Cannot use the SSID here: the hex-encode is done at build-time, not eval time
+    # and several special chars in a file name would break `writeText`.
+    file = pkgs.writeText "ssid" ''
+      [Security]
+      ${optionalString (cfg.psk != null) "PreSharedKey=${cfg.psk}"}
+      ${optionalString (cfg.passphrase != null) "Passphrase=${cfg.passphrase}"}
+
+      [Settings]
+      Autoconnect=${boolToString cfg.autoConnect}
+      Hidden=${boolToString cfg.hidden}
+    '';
+  });
 in {
-  options.networking.wireless.iwd.enable = mkEnableOption "iwd";
+  options.networking.wireless.iwd = {
+    enable = mkEnableOption "iwd";
+
+    interfaces = mkOption {
+      type = types.nullOr (types.listOf types.str);
+      example = [ "wlp2s0" ];
+      apply = x: if x == null then x else concatStringsSep " " x;
+      default = null;
+      description = ''
+        List of interfaces to use.
+      '';
+    };
```

Comment:
Can the interfaces change be done separately?
It's not needed for the network changes at all AFAICT so it's easily separated.
The reason is iwd and interface handling, especially naming, might need some careful work.

* Interface strings: interaction and reliance on use of `usePredictableInterfaceNames`
* Implications re:iwd's `UseDefaultInterface` (and the shipped `.link` file).
* (Is there some existing interface abstraction we should be tying to?)
* The service, I think?, if used with particular interfaces maybe should: `wants`, `requires` `sys-subsystem-net-devices-${ifacename-escaped-for-systemd}` (not sure)
* iwd's help doesn't tell you, but my reading of the code in `src/manager.c` suggests the interface list is comma-separated not space-separated, FWIW

All said, what's the use for specifying interfaces explicitly anyway? Just curious :).

And the rest is exciting and I think close to merge-ready, hence the suggestion for splitting off the interface-related bits.
Thoughts?

Reply:
> All said, what's the use for specifying interfaces explicitly anyway? Just curious :).

It might be possible that I'm missing something, but iwd constantly tried to use wlan0 by default which doesn't even exist on my system.

Reply:
@dtzWill any updates here?